compiler/rustc_codegen_ssa/src/back/archive.rs

```diff
@@ -28,32 +29,54 @@ pub trait ArchiveBuilderBuilder {
         is_direct_dependency: bool,
     ) -> PathBuf;
 
-    fn extract_bundled_libs(
-        &self,
-        rlib: &Path,
+    fn extract_bundled_libs<'a>(
+        &'a self,
+        rlib: &'a Path,
         outdir: &Path,
         bundled_lib_file_names: &FxHashSet<Symbol>,
-    ) -> Result<(), String> {
-        let message = |msg: &str, e: &dyn Display| format!("{} '{}': {}", msg, &rlib.display(), e);
+    ) -> Result<(), ExtractBundledLibsError<'_>> {
         let archive_map = unsafe {
-            Mmap::map(File::open(rlib).map_err(|e| message("failed to open file", &e))?)
-                .map_err(|e| message("failed to mmap file", &e))?
+            Mmap::map(File::open(rlib).map_err(|e| ExtractBundledLibsError {
+                kind: OpenFile,
+                rlib,
+                error: e.to_string(),
```

Comment:
I think we can use the `io::Error` in the type directly?

Reply:
We can't because different operations return different error types, here we have: `std::io::Error`, `object::Error` and `Utf8Error`.

We could accept a `&dyn Display` instead if you think that's better than `String` here.

Reply:
What about a `&dyn std::error::Error`?

Reply:
Updated [here](https://github.com/rust-lang/rust/pull/103792/commits/d43977c5fd7bac51bc22aa49bdc5d9cf9db52569) to `Box<dyn Error>` and made it conform to `IntoDiagnosticArg`. 
 
Please let me know if there is a better way to implement this.